src/System.Buffers.Experimental/System/Range.cs

```diff
@@ -1,26 +1,89 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-namespace System {
-    public struct Range
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System
+{
+
+    public struct Range : IEnumerable<int>
     {
-        public Range(long index, long length)
+        public const int UnboundedFirst = Int32.MinValue;
+        public const int UnboundedLast = Int32.MaxValue;
+        public const uint UnboundedLength = UInt32.MaxValue;
+
+        public readonly int First;
+        public readonly int Last; // Last is exclusive
+
+        public uint Length
+        {
+            get {
+                if (First == UnboundedFirst || Last == UnboundedLast) return UnboundedLength;
+                return (uint)((long)Last - (long)First);
+            }
+        }
+
+        public Range(int first, uint length)
         {
             if (length < 0)
             {
                 throw new ArgumentOutOfRangeException("Length must be non-negative.", nameof(length));
             }
 
-            Index = index;
-            Length = length;
+            First = first;
+            Last = (int)(first + length);
+
+            if (Last < First) throw new ArgumentOutOfRangeException(nameof(length));
+        }
+
+        private Range(int first, int last)
+        {
+            First = first;
+            Last = last;
+        }
+
+        public void Deconstruct(out int first, out int last)
+        {
+            first = First;
+            last = Last;
         }
-        public readonly long Index;
-        public readonly long Length;
+        public static Range Construct(int first, int last)
+            => new Range(first, last);
 
-        public void Deconstruct(out long index, out long length)
+        public struct Enumerator : IEnumerator<int>
         {
-            index = Index;
-            length = Length;
+            int _first;
+            int _last;
+
+            internal Enumerator(int first, int last)
+            {
+                _first = first - 1;
+                _last = last;
+            }
+
+            public bool MoveNext()
+            {
+                _first++;
```

Comment:
It seems that we're assuming that first < last, and therefore we should increment to go from `first` to `last`.
What if I create a `Range.Construct(10, 0)`?

I think such ranges should be allowed, but the enumerator should count down.

Reply:
I added TODO to consider.